### PR TITLE
Allow Iceberg MV with partitioning transforms on timestamptz

### PR DESCRIPTION
**Pull request by findepi**
Allow creation of Iceberg Materialized Views partitioned with a
temporarl partitioning function on a `timestamp with time zone` column.

In MVs, the `tunestano with time zone` columns are generally stored as
text to preserve time zone information. However, this prevents use of
temporal partitioning functions on these columns. The commit keeps
`timestamp with time zone` columns with partitioning applied on them as
`timestamp with time zone` in the storage table.

An obvious downside to this approach is that the time zone information
is erased and it is not known whether this aligns with user intention or
not. A better solution would be to introduce a point-in-time type
(https://github.com/trinodb/trino/issues/2273) to discern between the
cases where time zone information is important (like Java's
`ZonedDateTime`) from cases where only point-in-time matters (like
Java's `Instant`).

### CODE REVIEW

**Comment by findepi** [comment]
CI https://github.com/trinodb/trino/issues/13995, https://github.com/trinodb/trino/issues/16652 (new)

**Comment by findepi** [comment]
Build green (https://github.com/trinodb/trino/actions/runs/4481491874/jobs/7878270959?pr=16637).
Will force push for commit message fix.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 1]
```diff
@@ -217,9 +229,46 @@ protected SchemaTableName createMaterializedViewStorageTable(ConnectorSession se
...
+        Schema schemaWithTimestampTzPreserved = schemaFromMetadata(mappedCopy(
+                definition.getColumns(),
+                column -> {
+                    Type type = typeManager.getType(column.getType());
+                    if (type instanceof TimestampWithTimeZoneType timestampTzType && timestampTzType.getPrecision() <= 6) {
+                        // For now preserve timestamptz columns so that we can parse partitioning
+                        type = TIMESTAMP_TZ_MICROS;
+                    }
+                    else {
+                        type = typeForMaterializedViewStorageTable(type);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:242]: No need to update the TODO comment in `typeForMaterializedViewStorageTable` method? 
```
        if (type instanceof TimestampWithTimeZoneType) {
            // Iceberg does not store the time zone
            // TODO allow temporal partitioning on these columns, or MV property to
            //  drop zone info and use timestamptz directly
            return VARCHAR;
        }
```

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java [hunk 2]
```diff
@@ -594,13 +593,134 @@ public void testStorageSchemaProperty()
...
+    }
+
+    @Test(dataProvider = "testTemporalPartitioningDataProvider")
+    public void testTemporalPartitioning(String partitioning, String dataType, String exampleValue)
+    {
+        // validate the example value type
+        assertThat(query("SELECT " + exampleValue))
+                .matches("SELECT CAST(%s AS %S)".formatted(exampleValue, dataType));
+
+        assertUpdate("CREATE MATERIALIZED VIEW test_truncate_partitioning WITH (partitioning=ARRAY['%s(col)']) AS SELECT * FROM (VALUES CAST(NULL AS %s), %s) t(col)"
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:679]: thanks!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 3]
```diff
@@ -217,9 +229,46 @@ protected SchemaTableName createMaterializedViewStorageTable(ConnectorSession se
...
         SchemaTableName storageTable = new SchemaTableName(storageSchema, storageTableName);
-        List<ColumnMetadata> columns = definition.getColumns().stream()
-                .map(column -> new ColumnMetadata(column.getName(), typeForMaterializedViewStorageTable(typeManager.getType(column.getType()))))
-                .collect(toImmutableList());
+
+        Schema schemaWithTimestampTzPreserved = schemaFromMetadata(mappedCopy(
+                definition.getColumns(),
+                column -> {
+                    Type type = typeManager.getType(column.getType());
+                    if (type instanceof TimestampWithTimeZoneType timestampTzType && timestampTzType.getPrecision() <= 6) {
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:237]: It looks like we're going to use TIMESTAMP_TZ_MICROS for primitive type but if the same type is in a nested column then we're going to use VARCHAR.
Could you please add an explanation about why we need to do that ?

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:237]: we use varchar for all timestamp with time zone values, unless they are used for partitioning

Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:237]: Okay, so is this assuming that only non-nested types are allowed for partitioning ? Will this change when the other PR to allow partitioning on struct types lands ?

**Review by raunaqmorarka — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java [hunk 4]
```diff
@@ -594,13 +593,134 @@ public void testStorageSchemaProperty()
...
+        }
+    }
+
+    @DataProvider
+    public Object[][] testTemporalPartitioningDataProvider()
+    {
+        return new Object[][] {
+                {"year", "date", "DATE '2005-09-09'"},
+                {"year", "timestamp(6)", "TIMESTAMP '2005-09-10 13:31:00.123456'"},
+                {"year", "timestamp(6) with time zone", "TIMESTAMP '2005-09-10 13:00:00.123456 Europe/Warsaw'"},
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:700]: For the timestamp with time zone tests, can you add a second value that has the same time but in a different zone and show that works properly?

Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:700]: Also, a test with precision > 6?

Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:700]: > Also, a test with precision > 6?

this is not supported

**Pull request merged**